File: elixir/day11/lib/part1.ex
defmodule Part1 do
  def parse_input(input) do
    input
    |> String.split()
    |> Enum.map(&String.to_integer/1)
  end

  def process_stone(stone, 0), do: [stone]

  def process_stone(0, blinks), do: process_stone(1, blinks - 1)

  def process_stone(stone, blinks) do
    digits = Integer.to_string(abs(stone))
    process_stone_digits(stone, digits, blinks)
  end

  defp process_stone_digits(stone, digits, blinks) do
    case rem(String.length(digits), 2) do
      0 ->
        mid = div(String.length(digits), 2)
        {first_half, second_half} = String.split_at(digits, mid)

        [first_half, second_half]
        |> Enum.map(&String.to_integer/1)
        |> Enum.flat_map(&process_stone(&1, blinks - 1))

      _ ->
        process_stone(stone * 2024, blinks - 1)
    end
  end

  def blink(stones, blinks) do
    stones
    |> Enum.flat_map(&process_stone(&1, blinks))
    |> length()
  end

  def part1(input, blinks) do
    input
    |> parse_input()
    |> blink(blinks)
  end
end

File.read!("input.txt")
|> Part1.part1(25)
|> IO.inspect(label: "part1")

File: ruby/day11/part1.rb
# frozen_string_literal: true

def parse_input(input)
  input.split.map(&:to_i)
end

def process_stone(stone, blinks)
  return [stone] if blinks.zero?

  case stone
  when 0 then process_stone(1, blinks - 1)
  else
    digits = stone.abs.to_s
    if digits.length.even?
      mid = digits.length / 2
      [
        process_stone(digits[0...mid].to_i, blinks - 1),
        process_stone(digits[mid..].to_i, blinks - 1)
      ].flatten
    else
      process_stone(stone * 2024, blinks - 1)
    end
  end
end

def blink(stones, blinks)
  stones.flat_map { |stone| process_stone(stone, blinks) }.size
end

def part1(input, blinks)
  numbers = parse_input(input)
  blink(numbers, blinks)
end

if __FILE__ == $PROGRAM_NAME
  input = File.read('input.txt')
  puts part1(input, 25)
end
